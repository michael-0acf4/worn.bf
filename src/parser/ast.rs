#[derive(Debug, Clone)]
pub enum SuperValue {
    Integer(u32),
    String(String),
    Literal(String),
    SuperCall {
        callee: WithPos<String>,
        args: Vec<WithPos<Instruction>>,
    },
}

#[derive(Debug, Clone)]
pub enum Instruction {
    Add(i32),
    Move(i32),
    PutC,
    GetC,
    Loop {
        body: Vec<WithPos<Instruction>>,
    },
    // ext
    InlineValue(SuperValue),
    SuperFunction {
        name: WithPos<String>,
        args: Vec<WithPos<String>>,
        body: Vec<WithPos<Instruction>>,
    },
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct WithPos<T> {
    pub start: usize,
    pub end: usize,
    pub value: T,
}

impl<T> WithPos<T> {
    pub fn transfer<P>(&self, p: P) -> WithPos<P> {
        WithPos {
            value: p,
            start: self.start,
            end: self.end,
        }
    }
}

pub trait Reconstruct {
    fn reconstruct_at_depth(&self, depth: usize) -> String;

    fn reconstruct(&self) -> String {
        self.reconstruct_at_depth(0)
    }
}

impl<T: Reconstruct> Reconstruct for WithPos<T> {
    fn reconstruct_at_depth(&self, depth: usize) -> String {
        self.value.reconstruct_at_depth(depth)
    }
}

impl Reconstruct for SuperValue {
    fn reconstruct_at_depth(&self, depth: usize) -> String {
        let ret = match self {
            SuperValue::Integer(n) => format!("{}", n),
            SuperValue::String(s) => format!("{:?}", s),
            SuperValue::Literal(s) => format!("{s}"),
            SuperValue::SuperCall { callee, args } => {
                let arg_strs = args.iter().map(|arg| arg.reconstruct()).collect::<Vec<_>>();
                format!("{}({})", callee.value, arg_strs.join(", "))
            }
        };

        format!("{}{}", " ".repeat(depth), ret)
    }
}

impl Reconstruct for Instruction {
    fn reconstruct_at_depth(&self, depth: usize) -> String {
        match self {
            Instruction::Add(n) => {
                format!(
                    "{}{}",
                    " ".repeat(depth),
                    if *n > 0 {
                        "+".repeat(*n as usize)
                    } else {
                        "-".repeat((-n) as usize)
                    }
                )
            }
            Instruction::Move(n) => {
                format!(
                    "{}{}",
                    " ".repeat(depth),
                    if *n > 0 {
                        ">".repeat(*n as usize)
                    } else {
                        "<".repeat((-n) as usize)
                    }
                )
            }
            Instruction::Loop { body } => {
                let indent = " ".repeat(depth);
                format!(
                    "{}[\n{}\n{}]",
                    indent,
                    body.reconstruct_at_depth(depth + 1),
                    indent
                )
            }
            Instruction::PutC => format!("{}{}", " ".repeat(depth), '.'),
            Instruction::GetC => format!("{}{}", " ".repeat(depth), ','),
            Instruction::SuperFunction { name, args, body } => {
                let header = format!(
                    "super {}({})",
                    name.value,
                    args.iter()
                        .map(|a| a.value.clone())
                        .collect::<Vec<_>>()
                        .join(", ")
                );
                let body_str = body.reconstruct_at_depth(depth + 1);
                let indent = " ".repeat(depth);
                format!("{indent}{header} {{\n{body_str}\n{indent}}}\n")
            }
            Instruction::InlineValue(s) => s.reconstruct_at_depth(depth),
        }
    }
}

impl Reconstruct for Vec<WithPos<Instruction>> {
    fn reconstruct_at_depth(&self, depth: usize) -> String {
        self.iter()
            .map(|i| i.reconstruct_at_depth(depth))
            .collect::<Vec<_>>()
            .join("\n")
    }
}
