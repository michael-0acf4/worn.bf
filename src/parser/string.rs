//!
//! (Parsing unicode-safe Strings is hard! :'()
//! Modified/less-cryptic version built based on:
//! https://raw.githubusercontent.com/rust-bakery/nom/refs/heads/main/examples/string.rs

//!
//! A string is:
//!
//! - Enclosed by double quotes
//! - Can contain any raw unescaped code point besides \ and "
//! - Matches the following escape sequences: \b, \f, \n, \r, \t, \", \\, \/
//! - Matches code points like Rust: \u{XXXX}, where XXXX can be up to 6
//!   hex characters
//! - an escape followed by whitespace consumes all whitespace between the
//!   escape and the next non-whitespace character
//!

use super::Span;
use nom::{
    IResult,
    branch::alt,
    bytes::{complete::is_not, complete::take_while_m_n},
    character::complete::{char, multispace1},
    combinator::{map, map_opt, map_res, value, verify},
    multi::fold_many0,
    sequence::{delimited, preceded},
};
use nom_locate::LocatedSpan;

fn parse_unicode(input: Span) -> IResult<Span, char> {
    let parse_hex = take_while_m_n(1, 6, |c: char| c.is_ascii_hexdigit());
    let parse_delimited_hex = preceded(char('u'), delimited(char('{'), parse_hex, char('}')));
    let parse_u32 = map_res(parse_delimited_hex, |hex: Span| {
        u32::from_str_radix(hex.fragment(), 16)
    });

    map_opt(parse_u32, char::from_u32)(input)
}

fn parse_escaped_char(input: Span) -> IResult<Span, char> {
    preceded(
        char('\\'),
        alt((
            parse_unicode,
            value('\n', char('n')),
            value('\r', char('r')),
            value('\t', char('t')),
            value('\u{08}', char('b')),
            value('\u{0C}', char('f')),
            value('\\', char('\\')),
            value('/', char('/')),
            value('"', char('"')),
        )),
    )(input)
}

fn parse_escaped_whitespace(input: Span) -> IResult<Span, LocatedSpan<&str>> {
    preceded(char('\\'), multispace1)(input)
}

fn parse_literal(input: Span) -> IResult<Span, String> {
    let (input, fragment) = verify(is_not("\"\\"), |s: &Span| !s.fragment().is_empty())(input)?;
    Ok((input, fragment.fragment().to_string()))
}

#[derive(Debug, Clone, PartialEq, Eq)]
enum StringFragment {
    Literal(String),
    EscapedChar(char),
    EscapedWS,
}

fn parse_fragment(input: Span) -> IResult<Span, StringFragment> {
    alt((
        map(parse_literal, StringFragment::Literal),
        map(parse_escaped_char, StringFragment::EscapedChar),
        value(StringFragment::EscapedWS, parse_escaped_whitespace),
    ))(input)
}

pub fn parse_string(input: Span) -> IResult<Span, String> {
    // old:
    // let build_string = fold(0.., parse_fragment, String::new,...
    let build_string = fold_many0(parse_fragment, String::new, |mut string, fragment| {
        match fragment {
            StringFragment::Literal(s) => string.push_str(&s),
            StringFragment::EscapedChar(c) => string.push(c),
            StringFragment::EscapedWS => {}
        }
        string
    });

    delimited(char('"'), build_string, char('"'))(input)
}
