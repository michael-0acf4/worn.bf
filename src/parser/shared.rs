use super::ast::WithPos;
use nom::{
    IResult,
    branch::alt,
    bytes::complete::{tag, take_until},
    character::complete::{line_ending, multispace1},
    combinator::{opt, rest},
    multi::many0,
    sequence::{delimited, preceded, terminated},
};
use nom_locate::LocatedSpan;

pub type Span<'a> = LocatedSpan<&'a str>;

pub fn with_position_mut<'a, O>(
    mut parser: impl FnMut(Span<'a>) -> IResult<Span<'a>, O>,
) -> impl FnMut(Span<'a>) -> IResult<Span<'a>, WithPos<O>> {
    move |input: Span<'a>| {
        let start = input.location_offset();
        let (next_input, value) = parser(input)?;
        let end = next_input.location_offset();
        Ok((next_input, WithPos { value, start, end }))
    }
}

pub fn inline_comment(input: Span) -> IResult<Span, LocatedSpan<&str>> {
    preceded(
        tag("//"),
        // must terminate on \n or eof
        terminated(alt((take_until("\n"), rest)), opt(line_ending)),
    )(input)
}

pub fn multiline_comment(input: Span) -> IResult<Span, LocatedSpan<&str>> {
    delimited(tag("/*"), take_until("*/"), tag("*/"))(input)
}

/// 0 or more `space`, `inline_comment` or `multiline_comment`
pub fn skippable0(input: Span) -> IResult<Span, Vec<LocatedSpan<&str>>> {
    let noop = alt((inline_comment, multiline_comment, multispace1));
    many0(noop)(input)
}
