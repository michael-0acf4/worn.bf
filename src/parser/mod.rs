pub mod ast;
pub mod shared;
pub mod string;

use ast::{Instruction, SuperValue, WithPos};
use nom::{
    IResult,
    branch::alt,
    bytes::complete::{tag, take_while1},
    character::complete::char,
    combinator::map,
    multi::{many0, many1, separated_list0},
    sequence::{delimited, preceded},
};
use shared::{Span, skippable0, with_position_mut};
use string::parse_string;

fn parse_token(input: Span) -> IResult<Span, WithPos<String>> {
    map(
        with_position_mut(take_while1(|c: char| c.is_alphanumeric() || c == '_')),
        |s| s.transfer(s.value.to_string()),
    )(input)
}

fn parse_string_value(input: Span) -> IResult<Span, WithPos<SuperValue>> {
    map(with_position_mut(parse_string), |s| {
        s.transfer(SuperValue::String(s.value.clone()))
    })(input)
}

fn parse_literal_value(input: Span) -> IResult<Span, WithPos<SuperValue>> {
    map(parse_token, |s| {
        s.transfer(SuperValue::Literal(s.value.clone()))
    })(input)
}

fn parse_number_value(input: Span) -> IResult<Span, WithPos<SuperValue>> {
    map(
        with_position_mut(take_while1(|c: char| c.is_digit(10))),
        |s| {
            s.transfer(SuperValue::Integer(
                s.value.clone().to_owned().parse::<u32>().unwrap(),
            ))
        },
    )(input)
}

fn parse_value(input: Span) -> IResult<Span, WithPos<SuperValue>> {
    alt((
        parse_string_value,
        parse_super_call,
        parse_number_value,
        parse_literal_value,
    ))(input)
}

fn parse_super_call(input: Span) -> IResult<Span, WithPos<SuperValue>> {
    let start = input.location_offset();
    let (next_input, callee) = preceded(skippable0, parse_token)(input)?;
    let sep = delimited(skippable0, char(','), skippable0);
    let (next_input, args) = delimited(
        preceded(skippable0, char('(')),
        separated_list0(sep, preceded(skippable0, parse_instr)),
        preceded(skippable0, char(')')),
    )(next_input)?;

    Ok((
        next_input,
        WithPos {
            start,
            end: next_input.location_offset(),
            value: SuperValue::SuperCall { callee, args },
        },
    ))
}

fn parse_inline_value(input: Span) -> IResult<Span, WithPos<Instruction>> {
    map(parse_value, |v| {
        v.transfer(Instruction::InlineValue(v.value.clone()))
    })(input)
}

fn parse_add(input: Span) -> IResult<Span, WithPos<Instruction>> {
    map(
        with_position_mut(alt((many1(char('+')), many1(char('-'))))),
        |s| {
            s.transfer(Instruction::Add(
                s.value.len() as i32 * (if s.value[0] == '+' { 1 } else { -1 }),
            ))
        },
    )(input)
}

fn parse_loop(input: Span) -> IResult<Span, WithPos<Instruction>> {
    let body = delimited(
        preceded(skippable0, char('[')),
        many0(parse_instr),
        preceded(skippable0, char(']')),
    );

    map(with_position_mut(body), |s| {
        s.transfer(Instruction::Loop {
            body: s.value.clone(),
        })
    })(input)
}

fn parse_io(input: Span) -> IResult<Span, WithPos<Instruction>> {
    map(
        with_position_mut(alt((many1(char('.')), many1(char(','))))),
        |s| {
            let instr = if s.value[0] == '.' {
                Instruction::Put(s.value.len() as u32)
            } else {
                Instruction::Get(s.value.len() as u32)
            };

            s.transfer(instr)
        },
    )(input)
}

fn parse_move(input: Span) -> IResult<Span, WithPos<Instruction>> {
    map(
        with_position_mut(alt((many1(char('>')), many1(char('<'))))),
        |s| {
            s.transfer(Instruction::Move(
                s.value.len() as i32 * (if s.value[0] == '>' { 1 } else { -1 }),
            ))
        },
    )(input)
}

fn parse_super(input: Span) -> IResult<Span, WithPos<Instruction>> {
    let start = input.location_offset();
    let (next_input, _) = preceded(skippable0, tag("super"))(input)?;
    let (next_input, name) = preceded(skippable0, parse_token)(next_input)?;

    let sep = delimited(skippable0, char(','), skippable0);
    let (next_input, args) = delimited(
        preceded(skippable0, char('(')),
        separated_list0(sep, preceded(skippable0, parse_token)),
        preceded(skippable0, char(')')),
    )(next_input)?;

    let (next_input, body) = delimited(
        preceded(skippable0, char('{')),
        many1(parse_instr),
        preceded(skippable0, char('}')),
    )(next_input)?;

    Ok((
        next_input,
        WithPos {
            start,
            end: next_input.location_offset(),
            value: Instruction::SuperFunction { name, args, body },
        },
    ))
}

fn parse_instr(input: Span) -> IResult<Span, WithPos<Instruction>> {
    preceded(
        skippable0,
        alt((
            parse_add,
            parse_move,
            parse_io,
            parse_loop,
            parse_super,
            parse_inline_value,
        )),
    )(input)
}

pub fn parse_program(input: &str) -> Result<Vec<WithPos<Instruction>>, String> {
    let (_, instructions) = many0(parse_instr)(input.into()).map_err(|e| e.to_string())?;
    Ok(instructions)
}

#[test]
pub fn test_units() {
    // greedy test with take_till1
    assert_eq!(
        parse_token("a ".into()).map(|v| (v.0.to_string(), v.1.value)),
        Ok((" ".to_string(), "a".to_string()))
    );
    assert_eq!(
        parse_token("abc cd".into()).map(|v| (v.0.to_string(), v.1.value)),
        Ok((" cd".to_string(), "abc".to_string()))
    );
}
