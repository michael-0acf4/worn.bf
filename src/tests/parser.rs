use crate::parser::{ast::Reconstruct, parse_program};
use insta::assert_debug_snapshot;

#[test]
fn test_super_instruction() {
    assert_debug_snapshot!(
        parse_program(
            "
        super instr(arg1, arg2, arg3) {
            +-arg1 arg2 arg3
        }

        super instr2() {
            instr(++, --, super instr3() {
                instr(++, --, ++)
            })
        }
        ",
        )
        .map(|is| is.reconstruct())
    )
}

#[test]
fn test_expr_with_comments() {
    let out = parse_program(
        "
        /*
        * Awesome
        */

        super deluxe() { // skip me
        +
    }

        yay(1, 2, /* and me*/3,

        // and also me
        4,
        /* and me*/f(f(   5,   6 )))
        ",
    );
    assert_eq!(
        out.map(|is| is.reconstruct()),
        Ok("super deluxe() {\n +\n}\n\nyay(1, 2, 3, 4, f(f(5, 6)))".to_string())
    );
}
