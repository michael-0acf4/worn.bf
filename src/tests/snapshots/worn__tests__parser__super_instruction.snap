---
source: src/tests/parser.rs
expression: "parse_program(\"\n        super instr(arg1, arg2, arg3) {\n            +-arg1 arg2 arg3\n        }\n\n        super instr2() {\n            instr(++, --, super instr3() {\n                instr(++, --, ++)\n            })\n        }\n        \",).map(|is|\nis.reconstruct())"
snapshot_kind: text
---
Ok(
    "super instr(arg1, arg2, arg3) {\n +\n -\n arg1\n arg2\n arg3\n}\n\nsuper instr2() {\n instr(++, --, super instr3() {\n instr(++, --, ++)\n}\n)\n}\n",
)
